src/xdplwf/send.c

```diff
@@ -178,6 +178,130 @@ XdpGenericBuildTxNbl(
+                    Buffer->DataLength < sizeof(ETHERNET_HEADER) + (Ipv4->HeaderLength << 2) ||
+                    FrameLayout->Layer3HeaderLength != (Ipv4->HeaderLength << 2)) {
+                    goto InvalidOffload;
+                }
+
+                NextHeader = &Ipv4->Protocol;
+                ChecksumInfo->Transmit.IsIPv4 = TRUE;
+
+                break;
+
+            case XdpFrameLayer3TypeIPv6NoExtensions:
+            case XdpFrameLayer3TypeIPv6UnspecifiedExtensions:
+            case XdpFrameLayer3TypeIPv6WithExtensions:
+                const IPV6_HEADER *Ipv6 = RTL_PTR_ADD(Mdl->MappedSystemVa, sizeof(ETHERNET_HEADER));
+
+                if (Buffer->DataLength < sizeof(ETHERNET_HEADER) + sizeof(*Ipv6) ||
+                    Ipv6->Version != (IPV6_VERSION >> 4) ||
+                    FrameLayout->Layer3HeaderLength != sizeof(*Ipv6)) {
+                    goto InvalidOffload;
+                }
+
+                NextHeader = &Ipv6->NextHeader;
+                ChecksumInfo->Transmit.IsIPv6 = TRUE;
+
+                break;
+
+            default:
+                goto InvalidOffload;
+            }
+        }
+
+        if (FrameChecksum->Layer3 == XdpFrameTxChecksumActionRequired) {
+            switch (FrameLayout->Layer3Type) {
+            case XdpFrameLayer3TypeIPv4NoOptions:
+            case XdpFrameLayer3TypeIPv4UnspecifiedOptions:
+            case XdpFrameLayer3TypeIPv4WithOptions:
+                ChecksumInfo->Transmit.IpHeaderChecksum = TRUE;
+                break;
+            default:
+                goto InvalidOffload;
+            }
+        }
+
+        if (FrameChecksum->Layer4 == XdpFrameTxChecksumActionRequired) {
+            const UINT32 Layer4HeaderOffset =
+                FrameLayout->Layer2HeaderLength + FrameLayout->Layer3HeaderLength;
+            const VOID *Layer4Header = RTL_PTR_ADD(Mdl->MappedSystemVa, Layer4HeaderOffset);
+
+            switch (FrameLayout->Layer4Type) {
+            case XdpFrameLayer4TypeTcp:
+                const TCP_HDR *Tcp = Layer4Header;
+
+                if (*NextHeader != IPPROTO_TCP ||
+                    Buffer->DataLength < Layer4HeaderOffset + sizeof(*Tcp) ||
+                    Tcp->th_len < (sizeof(*Tcp) >> 2) ||
+                    Buffer->DataLength < Layer4HeaderOffset + (Tcp->th_len << 2)) {
+                    goto InvalidOffload;
+                }
+                ChecksumInfo->Transmit.TcpChecksum = TRUE;
+                ChecksumInfo->Transmit.TcpHeaderOffset = Layer4HeaderOffset;
+                break;
+
+            case XdpFrameLayer4TypeUdp:
+                const UDP_HDR *Udp = Layer4Header;
+
+                if (*NextHeader != IPPROTO_UDP ||
+                    Buffer->DataLength < Layer4HeaderOffset + sizeof(*Udp) ||
+                    Udp->uh_ulen < sizeof(*Udp) ||
+                    Buffer->DataLength < ntohs(Udp->uh_ulen) /* Allow UDP length to exceed IP length */) {
```

Comment:
Shouldn't it be the opposite? Can't IP length be more than the UDP length, because you can have that 'UDP option' space' after the UDP payload? But how could the UDP length be more than the IP?

Reply:
I'll clarify the comments, but the point of this check is simply to ensure the fields of each layer's headers fit within the memory buffer; it is trying to say that we don't attempt to validate that a packet would be an acceptable/RFC-conformant packet from L2 through L4.

Reply:
I added a similar comment for a similar non-check in TCP.